### PR TITLE
feat: add onceWriteDoc method to notify of indexed docs

### PR DESCRIPTION
**Pull request by sethvincent**
A simple approach for indicating when a doc has been indexed. This could then be emitted from a higher level module and used as confirmation that an edit has been made and indexed.

todo:
- [x] make sure this approach makes sense (revised approach)
- [x] add to api docs
- [x] fix types

### CODE REVIEW

**Comment by gmaclennan** [comment]
Looks good. Some questions:

1. Could it be an issue that this happens in the same tick / eventloop as the main code? It means that an expensive function for onWriteDoc will slow down the whole indexer. 
2. What is the overhead of calling a function for every doc (as opposed to running a check in the main body of code - see idea below) 

One possible variation of this is a kind of subscription model:

`indexer.onceWriteDoc({ seq, coreId }, (doc) => {})`

Would be called once, when the doc is written to the index. Could do this in the next tick if that makes sense. Somewhere the implementation is going to need some kind of "once" subscription: when writing to the db, write to the core then wait for the doc to pop out the other end at the indexer. 

I think the current API could work too though - part of it is just preference. You said you felt an event emitter would have overhead too? I wonder if we could quickly benchmark this - I feel like there are a lot of perf questions that we don't have answers to.

**Comment by sethvincent** [comment]
Yeah calling a function when indexing because of sync isn't ideal, so registering interest in a specific doc makes sense.

So far this module doesn't know about cores so I'm guessing you meant `docId` instead of `coreId`?

I just pushed a quick revision trying out a `onceWriteDoc` function and it seems good.

**Comment by sethvincent** [comment]
Should it be version instead of seq? I'm not sure which is easier to know / harder to be inaccurate.

**Comment by gmaclennan** [comment]
Yes sorry I forgot how this works. The `version` prop of documents in Mapeo is actually `${coreId}@${seq}` so we can use version to track a document through the stack. Eg indexer.onceWriteDoc(version, () => {})

**Review by gmaclennan — CHANGES_REQUESTED** [review]
Great work on this, pretty much there, just a couple of points:

1. It's currently only possible to add one listener per doc version (with the unexpected behaviour of "last listener wins"). Maybe make `this.#listeners` a Map of Sets. 
2. The listener is called with the flattened doc (eg links is a comma-separated string)

(2) would have been caught by declaring the type on the Map eg 

```
/** @type {Map<IndexCallback>} */
this.#listeners 
```

Regarding the flattened doc I think we should switch this module to using the SQLite json functions - we can do that in a follow up PR. I wrote this module before the json extension was included by default in SQLite (and before I properly understood it)

**Comment by sethvincent** [comment]
:+1: updated with those changes

File: index.js [hunk 1]
```diff
@@ -34,6 +34,9 @@ export class DbApi {
   #writeBacklinkSql
   #updateForksSql
   #docDefaults
+  /**
+   *
```
Comment (gmaclennan) [inline index.js]: ```@type {Map<string, Set<IndexCallback>>}``` (I think)

Comment (sethvincent) [reply index.js]: fixed

File: index.js [hunk 2]
```diff
@@ -94,6 +97,35 @@ export class DbApi {
...
+      process.nextTick(() => {
+        const set = this.#listeners.get(doc.version)
+        if (set) {
+          for (const listener of set.values()) {
+            listener(doc)
+          }
+          this.#listeners.delete(doc.version)
+        }
+      })
+    }
```
Comment (gmaclennan) [inline index.js:111]: This is just an abundance of caution (perhaps unnecessary):

Avoid failure if `doc` is mutated (e.g. version is changed) before the next tick by referencing the version directly (I know it's highly unlikely that we would do this, but if the code ended up doing this, then `doc.version` would refer to the mutated `version` in the next tick)

```suggestion
    const { version } = doc
    if (this.#listeners.has(version)) {
      process.nextTick(() => {
        const set = this.#listeners.get(version)
        if (set) {
          for (const listener of set.values()) {
            listener(doc)
          }
          this.#listeners.delete(version)
        }
      })
    }
```

File: index.js [hunk 3]
```diff
@@ -121,6 +153,7 @@ export class DbApi {
 
 export default class SqliteIndexer {
   #getWinner
+  #listeners = new Map()
```
Comment (gmaclennan) [inline index.js:156]: I think this is unused.

```suggestion
```

**Review by gmaclennan — APPROVED** [review]
Two small comments, non-blocking, good to merge.

File: index.js (hunk 3 above)
Comment (sethvincent) [reply index.js:157]: oops! thanks

**Pull request merged**